analyzer/evmtokens/evm_tokens.go

```diff
@@ -0,0 +1,219 @@
+// be careful of block analyzer potentially making further updates in between.
+// That's why, for example, there are separate last_mutate_round and
+// last_download_round columns. The block analyzer updates last_mutate_round,
+// this token analyzer updates last_download_round.
+
+const EmeraldDamaskTokensAnalyzerName = "emerald_damask_tokens"
+const MaxDownloadBatch = 20
+const DownloadTimeout = 61 * time.Second
+
+type Main struct {
+	source storage.RuntimeSourceStorage
+	target storage.TargetStorage
+	qf     analyzer.QueryFactory
+	logger *log.Logger
+}
+
+var _ analyzer.Analyzer = (*Main)(nil)
+
+func NewMain(cfg *config.AnalyzerConfig, target storage.TargetStorage, logger *log.Logger) (*Main, error) {
+	ctx := context.Background()
+
+	// Initialize source storage.
+	networkCfg := oasisConfig.Network{
+		ChainContext: cfg.ChainContext,
+		RPC:          cfg.RPC,
+	}
+	factory, err := oasis.NewClientFactory(ctx, &networkCfg, cfg.FastStartup)
+	if err != nil {
+		logger.Error("error creating client factory",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	network, err := analyzer.FromChainContext(cfg.ChainContext)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := analyzer.RuntimeEmerald.ID(network)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("Emerald runtime ID determined", "runtime_id", id)
+
+	client, err := factory.Runtime(id)
+	if err != nil {
+		logger.Error("error creating runtime client",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	return &Main{
+		source: client,
+		target: target,
+		qf:     analyzer.NewQueryFactory("oasis_3", analyzer.RuntimeEmerald.String()),
+		logger: logger.With("analyzer", EmeraldDamaskTokensAnalyzerName),
+	}, nil
+}
+
+type StaleToken struct {
+	Addr                  string
+	LastMutateRound       uint64
+	Type                  *modules.EVMTokenType
+	AddrContextIdentifier string
+	AddrContextVersion    int
+	AddrData              []byte
+}
+
+func (m Main) getStaleTokens(ctx context.Context, limit int) ([]*StaleToken, error) {
+	var staleTokens []*StaleToken
+	rows, err := m.target.Query(ctx, m.qf.RuntimeEVMTokensAnalysisStaleQuery(), limit)
+	if err != nil {
+		return nil, fmt.Errorf("querying discovered tokens: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var staleToken StaleToken
+		if err = rows.Scan(
+			&staleToken.Addr,
+			&staleToken.LastMutateRound,
+			&staleToken.Type,
+			&staleToken.AddrContextIdentifier,
+			&staleToken.AddrContextVersion,
+			&staleToken.AddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning discovered token: %w", err)
+		}
+		staleTokens = append(staleTokens, &staleToken)
+	}
+	return staleTokens, nil
+}
+
+func (m Main) processBatch(ctx context.Context) error {
+	staleTokens, err := m.getStaleTokens(ctx, MaxDownloadBatch)
+	if err != nil {
+		return fmt.Errorf("getting discovered tokens: %w", err)
+	}
+	m.logger.Info("processing", "num_stale_tokens", len(staleTokens))
```

Comment:
If `len(staleTokens)==0`, you can early exit here and avoid sending an empty batch to the DB

Reply:
let's add that

Reply:
added